File: 프로그래머스/3/43238. 입국심사/입국심사.java
class Solution {
    public long solution(int n, int[] times) {
        
        long r = 1;
        while (true) {
            if (n <= check(r, times))
                break;
            r *= 2;
        }
        long l = r / 10;
        long answer = 0;
        while (l <= r) {
            long mid = (l + r) / 2;
            
            if (check(mid, times) < n)
                l = mid + 1;
            else {
                answer = mid;
                r = mid - 1;
            }
        }
        if (n <= check(answer - 1, times))
            answer--;

        return answer;
    }
    
    public int check(long time, int[] times) {
        int answer = 0;
        for (int t : times) {
            answer += (time / t);
        }
        return answer;
    }
}

File: 프로그래머스/3/43238. 입국심사/입국심사.cpp
#include <string>
#include <vector>
#include <iostream>
using namespace std;

long long solution(int n, vector<int> times) {
    long long r = 1000000000000000000;
    long long l = 1, answer = r;
    
    while (l <= r) {
        long long mid = (l + r) / 2;
        long long sum = 0;
        for (int t : times)
            sum += mid / t;
        if (n <= sum) {
            answer = mid;
            r = mid - 1;
        }
        else
            l = mid + 1;     
    }
    return answer;
}